File: sdk/.net/ardvro.sdk.net/component/kenviro/controller/BackpropagationController.cs
namespace ardvro.sdk.net.component.kenviro.controller;

using ardvro.core.ext.connection;
using ardvro.core.lib.common;

public class BackpropagationController
{
    private IConnection _connection;
    public BackpropagationController(IConnection connection)
    {
        _connection = connection;
    }

    public void Learning(double[][] inputs, double[][]targets, double learningRate, double errorTarget, int epoch, BackpropagationSetting.BackpropagationSettingLayer[] layers, Action<BackpropagationSetting.BackprogationSettingResult> callback)
    {
        BackpropagationSetting setting = new BackpropagationSetting();
        setting.Epoch = epoch;
        setting.ErrorTarget = errorTarget;
        setting.LearningRate = learningRate;
        setting.Layers = layers;

        Flexible args = new Flexible();
        args.Add("Inputs", inputs);
        args.Add("Targets", targets);
        args.Add("Epoch", epoch);
        args.Add("Setting", setting);

        _connection.Submit(args, "ardvro/component/kenviro/Backpropagation/Learning", callback);
    }

    public void Compute(double[][] inputs, double[][][] weights, double[][]targets, double learningRate, double errorTarget, int epoch, BackpropagationSetting.BackpropagationSettingLayer[] layers, Action<double[]> callback)
    {
        BackpropagationSetting setting = new BackpropagationSetting();
        setting.Epoch = epoch;
        setting.ErrorTarget = errorTarget;
        setting.LearningRate = learningRate;
        setting.Layers = layers;

        Flexible args = new Flexible();
        args.Add("Inputs", inputs);
        args.Add("Weights", weights);
        args.Add("Targets", targets);
        args.Add("Epoch", epoch);
        args.Add("Setting", setting);

        _connection.Submit(args, "ardvro/component/kenviro/Backpropagation/Compute", callback);
    }

}

File: sample/kenviro/controller/BackpropagationController.js
var BackpropagationController = function BackpropagationController(cfg)
{
    let ctrl = {};

    ctrl.Learning = function (inputs, targets, learningRate, errorTarget, epoch, layers, callback)
    {
        let args = {
            Inputs: inputs,
            Targets: targets,
            Epoch: epoch,
            Setting: {
                LearningRate: learningRate,
                ErrorTarget: errorTarget,
                Layers: layers
            }
        };
        cfg.Connector.Submit(args, "ardvro/component/kenviro/Backpropagation/Learning", function (box)
        {
            callback != null ? callback(box == null ? null : box.Data) : null;
        });
    };

    ctrl.Compute = function (inputs, weights, biases, learningRate, errorTarget, epoch, layers, callback)
    {
        let args = {
            Inputs: inputs,
            Weights: weights,
            Biases: biases,
            Epoch: epoch,
            Setting: {
                LearningRate: learningRate,
                ErrorTarget: errorTarget,
                Layers: layers
            }
        };
        cfg.Connector.Submit(args, "ardvro/component/kenviro/Backpropagation/Compute", function (box)
        {
            callback != null ? callback(box == null ? null : box.Data) : null;
        });
    };


    return ctrl;
}